util/ribbon_impl.h

```diff
@@ -226,33 +285,71 @@ class StandardHasher {
     return bi.second;
   }
 
-  bool NextSeed(Seed max_seed) {
-    if (seed_ >= max_seed) {
-      return false;
-    } else {
-      ++seed_;
-      return true;
-    }
+  // Seed tracking APIs - see class comment
+  void SetRawSeed(Seed seed) { raw_seed_ = seed; }
+  Seed GetRawSeed() { return raw_seed_; }
+  void SetOrdinalSeed(Seed count) {
+    // A simple, reversible mixing of any size (whole bytes) up to 64 bits.
+    // This allows casting the raw seed to any smaller size we use for
+    // ordinal seeds without risk of duplicate raw seeds for unique ordinal
+    // seeds.
+
+    // Seed type might be smaller than numerical promotion size, but Hash
+    // should be at least that size, so we use Hash as intermediate type.
+    static_assert(sizeof(Seed) <= sizeof(Hash),
+                  "Hash must be at least size of Seed");
+
+    // Multiply by a large random prime (one-to-one for any prefix of bits)
+    Hash tmp = count * kToRawSeedFactor;
+    // Within-byte one-to-one mixing
+    static_assert((kMixMask & (kMixMask >> kMixShift)) == 0,
+                  "Illegal mask+shift");
+    tmp ^= (tmp & kMixMask) >> kMixShift;
+    raw_seed_ = static_cast<Seed>(tmp);
+    // dynamic verification
+    assert(GetOrdinalSeed() == count);
+  }
+  Seed GetOrdinalSeed() {
+    Hash tmp = raw_seed_;
+    // Within-byte one-to-one mixing (its own inverse)
+    tmp ^= (tmp & kMixMask) >> kMixShift;
+    // Multiply by 64-bit multiplicative inverse
+    static_assert(kToRawSeedFactor * kFromRawSeedFactor == Hash{1},
+                  "Must be inverses");
+    return static_cast<Seed>(tmp * kFromRawSeedFactor);
   }
-  Seed GetSeed() const { return seed_; }
-  void ResetSeed(Seed seed = 0) { seed_ = seed; }
 
  protected:
-  Seed seed_;
+  static constexpr Hash kMixMask = static_cast<Hash>(0xf0f0f0f0f0f0f0f0U);
+  static constexpr unsigned kMixShift = 4U;
+  static constexpr Hash kToRawSeedFactor =
+      static_cast<Hash>(0xc78219a23eeadd03U);
+  static constexpr Hash kFromRawSeedFactor =
+      static_cast<Hash>(0xfe1a137d14b475abU);
```

Comment:
Yes. If some number is congruent to 1 in mod 2^64, it's also congruent to 1 in mod 2^x for any whole x < 64 (since 2^x divides 2^64).

    >>> (0xadd03 * 0x475ab) & 0xfffff
    1